Microsoft.Toolkit.Uwp.UI.Controls.Input/RichSuggestBox/RichSuggestBox.cs

```diff
@@ -504,17 +545,25 @@ private async Task RequestSuggestionsAsync(ITextRange range = null)
                 return;
             }
 
-            CancelIfNotDisposed(this._suggestionRequestedCancellationSource);
-            this._suggestionRequestedCancellationSource = null;
+            var previousTokenSource = currentQuery?.CancellationTokenSource;
+            if (!(previousTokenSource?.IsCancellationRequested ?? true))
+            {
+                previousTokenSource.Cancel();
+            }
 
             if (queryFound)
             {
-                _currentQuery = new RichSuggestQuery { Prefix = prefix, QueryText = query, Range = range };
+                using var tokenSource = new CancellationTokenSource();
+                _currentQuery = new RichSuggestQuery
+                {
+                    Prefix = prefix,
+                    QueryText = query,
+                    Range = range,
+                    CancellationTokenSource = tokenSource
```

Comment:
Feel like the token shouldn't be exposed publicly here. I feel like we only should be calling Cancel if the previous operation (search) hasn't completed when another request has come in.

I don't see why we Cancel explicitly at the end of the method?

@azchohfi @Sergio0694 advice here? I think the general idea is that @huynhsontung wants to pass a token into the SuggestionsRequested event, so if the end-developer is doing a long-time search for results and the user types more into the box, that they can be notified to cancel the first search and start a new one. Right, that's the scenario?

Reply:
@michael-hawker `RichSuggestQuery` is a convenient internal structure that bundles all these properties together for internal tracking. It's not exposed to the public. `_currentQuery` is also a private field here so there is no chance of accessing it from outside. The token source is always canceled before disposed below (line 584).

Also, I have reworked the flow of this suggestion request a little bit to support something like `AdvancedCollectionView`. So you don't need to keep changing `ItemsSource` to update the suggestion list anymore. The new flow is: after the `SuggestionsRequested` event triggered, the suggestion popup always shows up unless the dev explicitly cancels it (through the event args) or if the suggestion list is empty. I hope that clears it up.

Reply:
Yes, that is correct! The idea is if the search is taking too long and the query is outdated, the older search should be canceled.

Reply:
@michael-hawker Sorry I missed the first part about why Canceling explicitly. So when the `CancellationTokenSource` is disposed, its `IsCancellationRequested` property is still accessible. To avoid synchronization problems between canceling and disposing, I just make sure all token sources are canceled before disposing. That way we can't cancel a disposed token source by checking for the `IsCancellationRequested` property first.

After reviewing the `InvokeAsync` code, the token is not exposed to the developer. In fact, there is no communication to the developer that the event is canceled. `InvokeAsync` would just stop waiting and move on. For this reason, I wonder if it is better to just skip cancellation on our end and let the developer handle the suggestion synchronization themselves. 🤔